### ISSUE TITLE
Unable to dump Profile info for the BERT model

### ISSUE THREAD

**Issue by Tengfei09** [NVIDIA/TensorRT#2003]
## Description

In TensorRT/demo/BERT, how do collect the timing results of all layers in bert_large_128.engine(like trtexec did)?

## Environment

**TensorRT Version**: 8.2.4
**NVIDIA GPU**: NVIDIA A100-PCIE(40GB)
**NVIDIA Driver Version**: 510.47
**CUDA Version**: 11.6
**CUDNN Version**: 8.2
**Operating System**: Ubuntu20.4
**Python Version (if applicable)**: 3.8
**Tensorflow Version (if applicable)**: 2.5
**PyTorch Version (if applicable)**: 
**Baremetal or Container (if so, version)**: **nvcr.io/nvidia/tensorrt:22.03-py3**

## Relevant Files

**TensorRT/demo/BERT**

## Steps To Reproduce

I just followed the quick start [guide](https://github.com/NVIDIA/TensorRT/tree/main/demo/BERT#quick-start-guide) provided by your team, to generate a **bert_large_128.engine**.

After that, I plan to dump the Profile information of all layers, as I know, there is a command to meet my requirement, 
`../tensorrt/bin/trtexec --loadEngine=engines/bert_large_128.engine --exportProfile=bert_profile.json`

However, I got an unexpected error as follows,

> [05/24/2022-12:48:34] [I] TensorRT version: 8.2.4
[05/24/2022-12:48:35] [I] [TRT] [MemUsageChange] Init CUDA: CPU +425, GPU +0, now: CPU 1076, GPU 685 (MiB)
[05/24/2022-12:48:35] [I] [TRT] Loaded engine size: 638 MiB
[05/24/2022-12:48:37] [I] [TRT] [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +1, GPU +8, now: CPU 2005, GPU 1895 (MiB)
[05/24/2022-12:48:37] [I] [TRT] [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +576, now: CPU 0, GPU 576 (MiB)
[05/24/2022-12:48:37] [I] Engine loaded in 2.71325 sec.
[05/24/2022-12:48:37] [I] [TRT] [MemUsageChange] Init cuBLAS/cuBLASLt: CPU +0, GPU +8, now: CPU 1366, GPU 1895 (MiB)
[05/24/2022-12:48:37] [I] [TRT] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +582, now: CPU 0, GPU 1158 (MiB)
[05/24/2022-12:48:37] [I] Using random values for input input_ids
[05/24/2022-12:48:37] [I] Created input binding for input_ids with dimensions 1x128
[05/24/2022-12:48:37] [I] Using random values for input segment_ids
[05/24/2022-12:48:37] [I] Created input binding for segment_ids with dimensions 1x128
[05/24/2022-12:48:37] [I] Using random values for input input_mask
[05/24/2022-12:48:37] [I] Created input binding for input_mask with dimensions 1x128
[05/24/2022-12:48:37] [I] Using random values for output cls_squad_logits
[05/24/2022-12:48:37] [I] Created output binding for cls_squad_logits with dimensions 1x128x2x1x1
[05/24/2022-12:48:37] [I] Starting inference
**Cuda failure: an illegal memory access was encountered
Aborted (core dumped)**

Then, I also failed to add IProfiler in TensorRT/demo/BERT/inference.py.

So, How to dump the timing results of all layers in BERT?

**Comment by zerollzeng** [comment]
I'm not very familiar with demoBERT, but dump profile is supported in build time, not inference time.

**Comment by Tengfei09** [comment]
> I'm not very familiar with demoBERT, but dump profile is supported in build time, not inference time.

No, According to TensorRT [docs](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#trtexec) ,  "--dumpProfile" is an option that used in the inference stage. it will print out  per-layer performance profile information.

> **Flags for the inference phase**
--loadEngine=<file>: Load the engine from a serialized plan file instead of building it from input ONNX, UFF, or Caffe model.
--batch=<N>: Specify the batch size to run the inference with. Only needed if the input models are in UFF or Caffe formats. If the input model is in ONNX format or if the engine is built with explicit batch dimension, use --shapes instead.
--shapes=<shapes>: Specify the input shapes to run the inference with.
--warmUp=<duration in ms>, --duration=<duration in seconds>, --iterations=<N>: Specify the minimum duration of the warm-up runs, the minimum duration for the inference runs, and the minimum iterations of the inference runs. For example, setting --warmUp=0 --duration=0 --iterations allows users to control exactly how many iterations to run the inference for.
--useCudaGraph: Capture the inference to a CUDA graph and run inference by launching the graph. This argument may be ignored when the built TensorRT engine contains operations that are not permitted under CUDA graph capture mode.
--noDataTransfers: Turn off host-to-device and device-to-host data transfers.
--streams=<N>: Run inference with multiple streams in parallel.
--verbose: Turn on verbose logging.
**--dumpProfile, --exportProfile=<file>: Print/Save the per-layer performance profile.**

**Comment by Tengfei09** [comment]
@nvpohanh If possible, could you help me to address this issue? More details have been shared before at this window.

**Comment by nvpohanh** [comment]
Assigning to myself. You're correct that dumpProfile is something at runtime.

**Comment by nvpohanh** [comment]
I will try to repro this first

**Comment by nvpohanh** [comment]
I can repro this, but this is not a bug in TensorRT. The issue is that `input_ids` and `segment_ids` are the token indices used to look up from the embedding tables. However, since trtexec uses random inputs, the embed lookup goes out-of-bound, causing cuda illegal memory access.

To solve this, you need to pass in "real input data" using trtexec's `--loadInputs` flag. Let me know if you need help on that.

**Comment by nvpohanh** [comment]
Or a simpler approach is to change the random range of the INT32 input from [-128, 127] to [0, 2]: https://github.com/NVIDIA/TensorRT/blob/main/samples/common/sampleUtils.h#L210

**Comment by Tengfei09** [comment]
> 
Thanks for your description, I get your point.
Currently, I don't have the "real input data" at hand. As for the alternative solution you provided, the only thing I need to do is change the range, do I need to recompile the _trtexec_?

**Comment by Tengfei09** [comment]
@nvpohanh  I could get what I want now.  Thank you very much!!!!

**Comment by nvpohanh** [comment]
@Tengfei09 Meanwhile, I am improving the embedding plugin used in demoBERT so that it doesn't do out-of-bound access even when the input indices are bad. Thanks for reporting this :)

**Comment by Tengfei09** [comment]
you mean _embLayerNormPlugin_ ? As I know, several optimized implementations of embedding have been provided by FasterTransformer. Is your implementation similar to theirs?

**Comment by nvpohanh** [comment]
Yes, that plugin. The issue is that these two lines are not doing correct boundary checking: https://github.com/NVIDIA/TensorRT/blob/main/plugin/embLayerNormPlugin/embLayerNormKernel.cu#L197-L198

And yes, this plugin is similar (if not almost identical) to that in FasterTransformers.
